expertise/service/expertise.py

```diff
@@ -0,0 +1,303 @@
+
+        # Validate + populate fields
+        for field in req_fields:
+            if field not in request:
+                raise OpenReviewException(f"Bad request: missing required field {field}")
+            config[field] = request[field]
+
+        for field in request.keys():
+            if field not in optional_fields and field not in req_fields:
+                raise OpenReviewException(f"Bad request: unexpected field {field}")
+            if field != 'model_params':
+                config[field] = request[field]
+
+        if 'model_params' in request.keys():
+            for field in request['model_params']:
+                if field not in optional_model_params:
+                    raise OpenReviewException(f"Bad request: unexpected model param: {field}")
+                config['model_params'][field] = request['model_params'][field]
+
+        # Populate with server-side fields
+        root_dir = os.path.join(self.working_dir, request['job_id'])
+        descriptions = JobDescription.VALS.value
+        config['dataset']['directory'] = root_dir
+        for field in path_fields:
+            config['model_params'][field] = root_dir
+        config['job_dir'] = root_dir
+        config['cdate'] = int(time.time())
+        config['status'] = JobStatus.INITIALIZED.value
+        config['description'] = descriptions[JobStatus.INITIALIZED]
+
+        # Set SPECTER+MFR paths
+        if config.get('model', 'specter+mfr') == 'specter+mfr':
+            config['model_params']['specter_dir'] = self.specter_dir
+            config['model_params']['mfr_feature_vocab_file'] = self.mfr_feature_vocab_file
+            config['model_params']['mfr_checkpoint_dir'] = self.mfr_checkpoint_dir
+
+        # Create directory and config file
+        if not os.path.isdir(config['dataset']['directory']):
+            os.makedirs(config['dataset']['directory'])
+        with open(os.path.join(root_dir, 'config.json'), 'w+') as f:
+            ## Remove the token before saving this in the file system
+            token = config.get('token', None)
+            if token is not None:
+                del config['token']
+                json.dump(config, f, ensure_ascii=False, indent=4)
+                config['token'] = token
+            else:
+                json.dump(config, f, ensure_ascii=False, indent=4)
+
+        return config
+
+    def _get_subdirs(self, user_id=None):
+        """
+        Returns the direct children directories of the given root directory
+
+        :returns: A list of subdirectories not prefixed by the given root directory
+        """
+        subdirs = [name for name in os.listdir(self.working_dir) if os.path.isdir(os.path.join(self.working_dir, name))]
+        if user_id is None:
+            return subdirs
+        else:
+            # If given a profile ID, assume looking for job dirs that contain a config with the
+            # matching profile id
+            filtered_dirs = []
+            for subdir in subdirs:
```

Comment:
is this scalable?

Reply:
@haroldrubio If we start using Redis to store the jobIDs and their major properties, we won't need to save directories for the jobs, right?